core/src/main/c/share/json.cpp

```diff
@@ -0,0 +1,592 @@
+    return simdjson::SIMDJSON_PADDING;
+}
+
+JNIEXPORT jstring JNICALL
+Java_io_questdb_std_json_SimdJsonError_errorMessage(
+        JNIEnv *env,
+        jclass /*cl*/,
+        jint code
+) {
+    auto msg = simdjson::error_message(static_cast<simdjson::error_code>(code));
+    return env->NewStringUTF(msg);
+}
+
+JNIEXPORT void JNICALL
+Java_io_questdb_std_json_SimdJsonParser_convertJsonPathToPointer(
+        JNIEnv * /*env*/,
+        jclass /*cl*/,
+        const char *json_chars,
+        size_t json_len,
+        questdb_byte_sink_t *dest_sink
+) {
+    const std::string_view json_path{json_chars, json_len};
+    const auto json_pointer = simdjson::ondemand::json_path_to_pointer_conversion(json_path);
+    auto dest = questdb_byte_sink_book(dest_sink, json_pointer.size());
+    std::memcpy(dest, json_pointer.data(), json_pointer.size());
+    dest_sink->ptr += json_pointer.size();
+}
+
+JNIEXPORT simdjson::ondemand::parser *JNICALL
+Java_io_questdb_std_json_SimdJsonParser_create(
+        JNIEnv * /*env*/,
+        jclass /*cl*/
+) {
+    return new simdjson::ondemand::parser();
+}
+
+JNIEXPORT void JNICALL
+Java_io_questdb_std_json_SimdJsonParser_destroy(
+        JNIEnv * /*env*/,
+        jclass /*cl*/,
+        simdjson::ondemand::parser *parser
+) {
+    delete parser;
+}
+
+JNIEXPORT void JNICALL
+Java_io_questdb_std_json_SimdJsonParser_queryPointerUtf8(
+        JNIEnv * /*env*/,
+        jclass /*cl*/,
+        simdjson::ondemand::parser *parser,
+        const char *json_chars,
+        size_t json_len,
+        size_t tail_padding,
+        const char *pointer_chars,
+        size_t pointer_len,
+        json_result *result,
+        questdb_byte_sink_t *dest_sink,
+        int32_t max_size
+) {
+    value_at_pointer(
+            parser, json_chars, json_len, tail_padding, pointer_chars, pointer_len, result,
+            [result, dest_sink, max_size](simdjson_value res) -> token_void {
+                switch (result->type) {
+                    case simdjson::ondemand::json_type::string:
+                        truncated_utf8_copy(
+                                *dest_sink,
+                                res.get_string().value_unsafe(),
+                                static_cast<size_t>(max_size));
+                        return {};
+                    case simdjson::ondemand::json_type::array:
+                    case simdjson::ondemand::json_type::object:
+                    case simdjson::ondemand::json_type::number:
+                    case simdjson::ondemand::json_type::boolean:
+                        extract_raw_json(res, dest_sink, max_size, *result);
+                        return {};
+                    case simdjson::ondemand::json_type::null:
+                        return default_value<token_void>::value();
+                }
+                return default_value<token_void>::value();
+            });
+}
+
+JNIEXPORT jlong JNICALL
+Java_io_questdb_std_json_SimdJsonParser_queryPointerValue(
```

Comment:
Why not keep this method only and move all type-casting to the Java side?

Reply:
Two reasons:
* On invalid input (e.g. parsing dobules when the doc has strings) it's faster to fail in C++ without returning and populating a throwaway buffer.
* The `queryPointerValue` API was actually added last and it's intended for DATE, IPV4 and TIMESTAMP parsing only.